src/cocoa/toga_cocoa/widgets/table.py

```diff
@@ -37,8 +37,17 @@ def tableView_objectValueForTableColumn_row_(self, table, column, row: int):
 
         col_identifier = str(column.identifier)
 
-        datum = data[col_identifier]
-        value = getattr(data_row, col_identifier)
+        # obtain data from 'col_identifier'
+        # and constructing if doesn't exists
+        try:
+            datum = data[col_identifier]
+        except KeyError:
+            data[col_identifier] = TogaData.alloc().init()
+            data_row._impl[col_identifier] = data[col_identifier]
+            datum = data[col_identifier]
+
+        # empty default value to avoid crash when a column is added
+        value = getattr(data_row, col_identifier, '')
```

Comment:
How does this case manifest? Is it an issue with the sequencing of events where the tree refreshes *before* the data exists to support it? If that's the case, a better approach would to fix the underlying ordering problem.

Reply:
If we don't put a default empty value it will crash when the datasource doesn't has the column. Maybe I can explain better with examples than with words:

### This will crash
```python
datasource = [
    { 'column_a': '1', 'column_b': '2'}
]
self.table = toga.Table(headings=['Column 1', 'Column 2'], accessors=['column_a', 'column_b'])
self.table.data = datasource
...
self.table.add_column('Column 3', 'column_c') # Crash because datasource does not have column_c
```

### This will work ok
```python
datasource = [
    { 'column_a': '1', 'column_b': '2', 'column_c': '3'}
]
self.table = toga.Table(headings=['Column 1', 'Column 2'], accessors=['column_a', 'column_b'])
self.table.data = datasource
...
self.table.add_column('Column 3', 'column_c')
```

Maybe the solution for this is check if data has `column_c`. If data doesn't has that column, don't allow to create it. Do you think we have to resolve in this direction?

Reply:
That makes a lot more sense - thanks for the explanation. 

I like the *idea* of preventing a column from being added unless the accessor can be satisfied; however, I don't think that solution will work across the board. For example, if your data is:
```
datasource = {
    { 'column_a': '1', 'column_b': '2', 'column_c': '3'},
    { 'column_a': '4', 'column_b': '5'},
]
```
does this data source satisfy the accessor or not? 

If the answer is "not" - what about if there were 10k rows in the table? How do we determine if the column name is valid without scanning the entire data table?

Or - what about if you created a table as you describe in your second example, and then you add `{ 'column_a': '4', 'column_b': '5'}` to your data? The table *was* valid, but now it isn't. So we either have to raise an error on the call to add the row, or drop the column.

My suggestion would be to add an option to Table that effectively tracks "how do you want to handle missing data?". The "fallback to empty string" behavior you've got in this patch would essentially be the equivalent of `table.missing_value = ''`. But, if you wanted to, you could use `table.missing_value = '???'` to use a different string; and, as a special case, `table.missing_value = None` would output a message to console as a warning, and fall back to '' as a placeholder value. I'd argue "log an error" would be the best *default* behavior, with a log message like:
`Row {n} of table data doesn't support accessor `{accessor}`. Using empty string; define a 'missing_value' on the table to silence this message`.

Does that make sense? Can you think of a better alternative?

Reply:
Makes all sense, I was thinking I did't figure out a better alternative so I implemented it as you said